src/webrtc/gcc.md

```diff
@@ -0,0 +1,159 @@
+to see if the network can handle more data.
+
+Unfortuantely, the connection probing is not standardized so every SFU might do this in a different way.
+
+The questions we need to answer are:
+1. what data use for probing
+1. how much to probe
+1. how often to probe
+1. how long to probe
+
+### 1. What data use for probing
+
+The naive implementation would be to switch to the higher layer every `x` seconds and observe the network.
+This however, may lead to periodic video rebuffering and lower Quality of Experience (QoE).
+
+Another solution is to generate traffic by sending RTP padding packets.
+In this case, server has full control over the amout of data being sent and therefore, it can probe with 
+different aggresivness.
+
+The problem with padding packets is that they contain only zeros (except the last byte) so we end up sending
+a lot of garbage data, useless for the client. 
+Fortunatelly, there is a solution which is called RTX.
+RTX is a shortcut for retransmissions and can be used for probing instead of padding packets.
+How does it work?
+Instead of sending padding packets we simply retranssmit some packets via dedicated RTP stream.
+This way we have a full control over probing rate and we send data that is useful for the client.
+That's also the way libwebrtc chooses when RTX extension is enabled.
+
+#### Probing with padding packets
+
+In our webrtc implementation, we started with padding packets as they are easier for implementing and give
+us pretty good results.
+Here are some tips important when using padding packets:
+* the `P` bit in the RTP header has to be set to `1`.
+It indicates that RTP packet contains padding at the end of a payload.  
+* maximal size of padding only payload is 255.
+The limitation comes from the last byte. When the `P` flag is set, the last byte in payload denotes number of 
+zeros that should be stripped out.
+* padding packets can only be sent at video frame boundary.
+In other case the decoder in a web browser might get out-of-sync and you will see a lot of freezes.
+How can you know when a video frames ends?
+Just check the `M` bit in the RTP header.
+For most codecs, `M` bit set to `1` means the end of a video frame.
+* you have to take care of rewriting sequence numbers
+* as a timestamp for padding packet you can use the timestamp of the last RTP packet with `M` bit set
+* sending padding packets will cause your SRTP window to move forward. 
+This might be problematic for out-of-order packets making it impossible to encrypt them as their sequence number
+will not fit into the encryption window.
+To solve this, you can either carefully increase the encryption window (too big encryption window might
+introduce security vulnerabilities) or just drop the packet.
+
+### 2. How much to probe
+
+The estimation can increase maximum by 8% per second.
+This means that there is no sense in probing much more than your estimation can increase.
```

Comment:
Nice point. Shouldn't we then rewrite some of our probing targets to target, let's say, 110% of GCC estimation?

Reply:
Sounds good, I am curious the results. In particular whether there will be any performance gains :thinking: